### PR TITLE
[Feature] Institution Login for OKState University and Delegate Authentication to External CAS Server [#CAS-15][#OSF-7271]

### PR DESCRIPTION
**Pull request by cslzchen**
### Deployment Notes

- Model Migration Required for `osf_institution`.
- Populate Institution Required.

### Tickets and PRs

- Replaces #6689 and #6687 
- CAS PR: https://github.com/CenterForOpenScience/cas-overlay/pull/42
- CAS Ticket: https://openscience.atlassian.net/browse/CAS-15
- OSF Ticket: https://openscience.atlassian.net/browse/OSF-7271

### CODE REVIEW

File: framework/auth/cas.py [hunk 1]
```diff
@@ -285,7 +285,7 @@ def make_response_from_ticket(ticket, service_url):
             # if user is authenticated by CAS
             return authenticate(
                 user,
-                cas_resp.attributes['accessToken'],
+                cas_resp.attributes.get('accessToken', ''),
```
Comment (cslzchen) [inline framework/auth/cas.py]: Set `accessToken` to empty if service validation does not release it in attributes. Please refer to this ticket: https://openscience.atlassian.net/browse/CAS-27.

File: scripts/populate_institutions.py [hunk 2]
```diff
@@ -137,17 +144,19 @@ def main(env):
...
             },
             {
                 '_id': 'okstate',
                 'name': 'Oklahoma State University',
                 'description': '<a href="http://www.library.okstate.edu/research-support/research-data-services/">OSU Library Research Data Services</a>',
                 'banner_name': 'okstate-banner.png',
                 'logo_name': 'okstate-shield.png',
-                'login_url': None,  # https://stwcas.okstate.edu/cas/login?service=...
-                'logout_url': None,
+                'login_url': None,
```
Comment (cslzchen) [inline scripts/populate_institutions.py]: `cas-pac4j` based institution login has a `None` login url but uses the default `saml-shib` logout url.

File: framework/auth/__init__.py [hunk 3]
```diff
@@ -125,7 +125,7 @@ def get_or_create_user(fullname, address, reset_password=True, is_spam=False):
     else:
         password = str(uuid.uuid4())
         user = User.create_confirmed(address, password, fullname)
-        if password:
+        if reset_password:
```
Comment (cslzchen) [inline framework/auth/__init__.py]: From: https://github.com/CenterForOpenScience/osf.io/pull/6497

File: api/institutions/authentication.py [hunk 4]
```diff
@@ -27,42 +55,45 @@ def authenticate(self, request):
...
 
-        username = provider['user']['username']
-        fullname = provider['user']['fullname']
+        username = provider['user'].get('username')
+        fullname = provider['user'].get('fullname')
+        given_name = provider['user'].get('givenName')
+        family_name = provider['user'].get('familyName')
+
+        # use given name and family name to build full name if not provided
+        if given_name and family_name and not fullname:
```
Comment (cslzchen) [inline api/institutions/authentication.py]: OSF requires `fullname` when creating a user. If `fullname` is not provided:
1. construct `fullname` from `given_name` and `family_name`

File: api/institutions/authentication.py [hunk 5]
```diff
@@ -27,42 +55,45 @@ def authenticate(self, request):
...
+            fullname = given_name + ' ' + family_name
+
+        # use username if no names are provided
+        if not fullname:
```
Comment (cslzchen) [inline api/institutions/authentication.py]: 2. if the above construction failed, use `username` as `fullname`

File: api/institutions/authentication.py [hunk 6]
```diff
@@ -27,42 +55,45 @@ def authenticate(self, request):
...
+            fullname = username
 
         user, created = get_or_create_user(fullname, username, reset_password=False)
 
         if created:
-            user.given_name = provider['user'].get('givenName')
+            # `get_or_create_user()` guesses given name and family name from fullname
+            # replace the guessed ones if the names are provided from the authentication request
+            if given_name:
```
Comment (cslzchen) [inline api/institutions/authentication.py]: Replace OSF guessed names with the names provided from the institution.

File: osf/migrations/0007_auto_20170404_1415.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    dependencies = [
+        ('osf', '0006_add_jsonb_index_for_fileversions'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='institution',
+            name='delegation_protocol',
+            field=models.CharField(blank=True, choices=[(b'cas-pac4j', b'CAS by pac4j'), (b'oauth-pac4j', b'OAuth by pac4j'), (b'saml-shib', b'SAML by shibboleth'), (b'', b'No Delegation Protocol')], default=b'', max_length=15),
```
Comment (icereval) [inline osf/migrations/0007_auto_20170404_1415.py]: choices should follow this implementation https://github.com/CenterForOpenScience/osf.io/blob/084bc28dbf56d614451659655c093ff745850869/osf/models/sanctions.py#L34

File: osf/models/institution.py [hunk 8]
```diff
@@ -17,19 +17,42 @@ class Institution(Loggable, base.ObjectIDMixin, base.BaseModel):
...
+    logout_url = models.URLField(null=True, blank=True)
+
     domains = fields.ArrayField(models.CharField(max_length=255), db_index=True, null=True, blank=True)
     email_domains = fields.ArrayField(models.CharField(max_length=255), db_index=True, null=True, blank=True)
-    logo_name = models.CharField(max_length=255, null=True)  # TODO: Could this be a FilePathField?
-    logout_url = models.URLField(null=True, blank=True)
-    name = models.CharField(max_length=255)
 
-    description = models.TextField(blank=True, default='', null=True)
+    contributors = models.ManyToManyField(
```
Comment (icereval) [inline osf/models/institution.py]: change is unrelated to the pr

File: osf/models/institution.py [hunk 9]
```diff
@@ -17,19 +17,42 @@ class Institution(Loggable, base.ObjectIDMixin, base.BaseModel):
...
+    # The protocol used to delegate authentication: `CAS`, `SAML`, `OAuth`, e.t.c
+    # We use shibbloeth's implementation for SAML and pac4j's implementation for CAS and OAuth
+    # Only institutions with a valid delegation protocol shows up on institution login page
+    DELEGATION_PROTOCOL_CHOICES = (
+        ('cas-pac4j', 'CAS by pac4j'),
+        ('oauth-pac4j', 'OAuth by pac4j'),
+        ('saml-shib', 'SAML by shibboleth'),
+        ('', 'No Delegation Protocol'),
+    )
+    delegation_protocol = models.CharField(max_length=15, choices=DELEGATION_PROTOCOL_CHOICES, blank=True, default='')
```
Comment (icereval) [inline osf/models/institution.py]: `blank=True` == `(default='' and null=False)`

File: framework/auth/cas.py (hunk 1 above)
Comment (icereval) [inline framework/auth/cas.py]: please add TOOD to cleanup

File: framework/auth/cas.py [hunk 10]
```diff
@@ -299,7 +299,7 @@ def make_response_from_ticket(ticket, service_url):
                 'external_id_provider': external_credential['provider'],
                 'external_id': external_credential['id'],
                 'fullname': fullname,
-                'access_token': cas_resp.attributes['accessToken'],
+                'access_token': cas_resp.attributes.get('accessToken', ''),
```
Comment (icereval) [inline framework/auth/cas.py]: same ^

File: api/institutions/authentication.py (hunk 5 above)
Comment (icereval) [inline api/institutions/authentication.py]: expose their email address?

Comment (icereval) [inline api/institutions/authentication.py]: at this point just fail, the idp can do better

File: api/institutions/authentication.py [hunk 11]
```diff
@@ -26,42 +54,50 @@ def authenticate(self, request):
...
+        middle_names = provider['user'].get('middleNames')
+        suffix = provider['user'].get('suffix')
 
-        user, created = get_or_create_user(fullname, username, reset_password=False)
+        # use given name and family name to build full name if not provided
+        if given_name and family_name and not fullname:
+            fullname = given_name + ' ' + family_name
+
+        # institution must provide `fullname`, otherwise we fail the authentication and inform sentry
+        if not fullname:
```
Comment (cslzchen) [inline api/institutions/authentication.py]: Inform Sentry and raise `AuthenticationFailed` exception if `fullname` is not provided or cannot be constructed from `given_name` and `family_name`.

File: framework/auth/cas.py [hunk 12]
```diff
@@ -283,9 +283,10 @@ def make_response_from_ticket(ticket, service_url):
                 )))
 
             # if user is authenticated by CAS
+            # TODO [CAS-27]: Remove Access Token From Service Validation
```
Comment (cslzchen) [inline framework/auth/cas.py]: Add a TODO with CAS Ticket for future clean up.

File: framework/auth/cas.py [hunk 13]
```diff
@@ -295,11 +296,12 @@ def make_response_from_ticket(ticket, service_url):
             fullname = u'{} {}'.format(cas_resp.attributes.get('given-names', ''), cas_resp.attributes.get('family-name', '')).strip()
             if not fullname:
                 fullname = external_credential['id']
+            # TODO [CAS-27]: Remove Access Token From Service Validation
```
Comment (cslzchen) [inline framework/auth/cas.py]: Same as above.

File: osf/models/institution.py [hunk 14]
```diff
@@ -8,19 +8,43 @@
...
+    # TODO Could `banner_name` and `logo_name` be a FilePathField?
+    banner_name = models.CharField(max_length=255, blank=True, null=True)
+    logo_name = models.CharField(max_length=255, blank=True, null=True)
+
+    # The protocol which is used to delegate authentication.
+    # Currently, we have `CAS`, `SAML`, `OAuth` available.
+    # For `SAML`, we use Shibboleth.
+    # For `CAS` and `OAuth`, we use pac4j.
+    # Only institutions with a valid delegation protocol show up on the institution login page.
+    DELEGATION_PROTOCOL_CHOICES = (
```
Comment (cslzchen) [inline osf/models/institution.py]: Update the value to be consistent with Java (CAS-side) `enum`.

**Comment by cslzchen** [comment]
Change target from `develop` to `master`. Failed to rebase or cherry-pick. Closed and replaced by #7120.
